rfd/0032-role-tester.md

```diff
@@ -0,0 +1,94 @@
+
+### Rules
+Rules are sentences that allow us to infer new facts from existing ones. We can combine multiple rules to create new rules. The most important question for the role tester is whether a user can access a given node as an os user, and we will define this as HasAccess(User, Login, Node). Other rules will provide other contextually related facts that will help answer this overarching question.
+
+Rule | Logical interpretation
+--- | ---
+HasAllowedSSHLabel(Role, Node, Key, Value) :- RoleAllowedSSHLabel(Role, Key, Value), NodeHasSSHLabel(Node, Key, Value) | If the role has allow labels and the node has the same labels, then the role allows access to the node
+HasDeniedSSHLabel(Role, Node, Key, Value) :- RoleDeniedSSHLabel(Role, Key, Value), NodeHasSSHLabel(Node, Key, Value) | If the role has deny labels and the node has the same labels, then the role denies access to the node
+HasRolePath(User, Login, Node) :- HasRole(User, Role), HasAllowedSSHLabel(Role, Node, Key, Value), RoleHasSSHLogin(Role, Login) | If the user has a role and the role allows access to the node and the role has a login, then the user has a role that gives them access to the node as login
+HasDeniedRolePath(User, Node) :- HasRole(User, Role), HasDeniedSSHLabel(Role, Node, Key, Value) | If the user has a role and the role denies access to the node, then the user has a role that denies them access to the node
+HasAccess(User, Login, Node) :- HasRolePath(User, Login, Node), not(HasDeniedRolePath(User, Node)) | If the user has a role that gives them access to the node as login and there are no roles denying access to the node, then the user has access to the node as login
+RoleAllowed(User, Role, Node) :- HasRole(User, Role), HasAllowedSSHLabel(Role, Node, Key, Value) | If the user has a role and the role allows access to the node, then that specific role gives the user access to node
+RoleDenied(User, Role, Node) :- HasRole(User, Role), HasDeniedSSHLabel(Role, Node, Key, Value) | If the user has a role and the role denies access to the node, then that specific role denies the user access to node
+
+Rules can be thought of as a simple implication between the body (the part after :-) and the head (the part before :-) where if body is true, then it implies the head.
+
+### Examples
+Some common questions are listed below. It is important to note that if we provide variables instead of constants to the queries, we will get a list of term values for the variable based on the facts that exist or are inferred. For example, HasAccess(jean, root, Node)? will return all the SSH nodes that user 'jean' has access to as 'root'. Similarly, the queries RoleAllowed? and RoleDenied? can be used in this way to get the corresponding roles that deny/allow a user access to a specific SSH node.
+
+Example | Query interpretation
+--- | ---
+HasAllowedSSHLabel(dev, node-1, environment, staging)? | Does the role 'dev' allow access to SSH node 'node-1' given the label 'environment:staging'?
```

Comment:
style nit on queries: I wonder if we can make predicates shorter for shorter queries, typing 'HasDeniedRolePath' in repl could take a while :)

Reply:
Agreed. I have modified the naming slightly so it's shorter but still self-descriptive, but open to better naming schemes.